Update native hosting doc with component loading specifics (#6680)
dotnet/core-setup: Documentation/design-docs/native-hosting.md

The library would be a dynamically loaded library (`.dll`, `.so`, `.dylib`). For ease of use there would be a header file for C/C++ apps as well as `.lib`/`.a` for easy linking.
Native host would ship this library as part of the app. Unlike the `apphost`, `comhost` and `ijwhost`, the `nethost` will not be directly supported by the .NET Core SDK since it's target usage is not from .NET Core apps.

> The `nethost` is part of the `Microsoft.NETCore.DotNetAppHost` package. Users are expected to either download the package directly or rely on .NET SDK to pull it.

Reviewer: ```suggestion
The `nethost` is part of the `Microsoft.NETCore.DotNetAppHost` package. Users are expected to either download the package directly or rely on .NET SDK to pull it down.
```

Starts the runtime and returns a function pointer to specified functionality of the runtime.
* `host_context_handle` - handle to the initialized host context.
* `type` - the type of runtime functionality requested
  * `hdt_load_assembly_and_get_function_pointer` - entry point which loads an assembly (with dependencies) and returns function pointer for a specified static method. See below for details (Loading managed components)

>   * `hdt_com_activation`, `hdt_com_register`, `hdt_com_unregister` - COM activation entry-point - see [COM activation](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/COM-activation.md) for more details.

Reviewer 2: ```suggestion
  * `hdt_com_activation`, `hdt_com_register`, `hdt_com_unregister` - COM activation entry-points - see [COM activation](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/COM-activation.md) for more details.
```

* `host_context_handle` - handle to the initialized host context to close.

### Loading managed components
To load managed components from native app directly (not using COM or WinRT) the hosting layer exposes a new runtime helper/delegate `hdt_load_assembly_and_get_function_pointer`. Calling the `hostfxr_get_runtime_delegate(handle, hdt_load_assembly_and_get_function_pointer, &helper)` returns a function pointer to the runtime helper with this signature:
```C

> int load_assembly_and_get_function_pointer_fn(

Reviewer 2: Do we want to talk about x86 calling convention?

Author: I'm intentionally leaving out the technical details - I think it doesn't belong to this document. Ideally we should have a header file which defines that in programmatic way which anybody can understand perfectly.

    const char_t *assembly_path,
    const char_t *type_name,
    const char_t *method_name,
    const char_t *delegate_type_name,
    void         *reserved,
    /*out*/ void **delegate)
```

> Calling this function will load the specified assembly in isolation (into its own `AssemblyLoadContext`) and it will use `AssemblyDependencyResolver` on it to provide dependency resolution. Once loaded it will find the specified type and method and return a native function pointer to that method. The method's signature can be specified via the delegate type name.

Reviewer 2: Do we want to talk about reuse using `assembly_path` as a dictionary key?

Author: I thought about that. I added a description of this and why components should not rely on it,,,

* `assembly_path` - Path to the assembly to load. In case of complex component, this should be the main assembly of the component (the one with the `.deps.json` next to it). Note that this does not have to be the assembly from which the `type_name` and `method_name` are.
* `type_name` - Assembly qualified type name to find
* `method_name` - Name of the method on the `type_name` to find. The method must be `public static` and must match the signature of `delegate_type_name`.
* `delegate_type_name` - Assembly qualified delegate type name for the method signature, or null. If this is null, the method signature is assumed to be:
```C#
public delegate int ComponentEntryPoint(IntPtr args, int sizeBytes);
```
This maps to native signature:
```C
int component_entry_point_fn(void *arg, int32_t arg_size_in_bytes);
```
* `reserved` - parameter reserved for future extensibility, currently unused and must be `0`.
* `delegate` - out parameter which receives the native function pointer to the requested managed method.

> It is allowed to call the returned runtime helper many times for different assemblies or just different methods from the same assembly. It is not required to get the helper every time. The implementation of the helper will cache loaded assemblies, so asking to load the same assembly twice will end up loading it only once and reusing it from that point onward. Ideally components should not take a dependency on this behavior, which means components should not have global state. Global state in components is typically just cause for problems. For example it may create ordering issues or unintended side effects and so on.

Reviewer: ```suggestion
It is allowed to call the returned runtime helper many times for different assemblies or different methods from the same assembly. It is not required to get the helper every time. The implementation of the helper will cache loaded assemblies, so requests to load the same assembly twice will load it only once and reuse it from that point onward. Ideally components should not take a dependency on this behavior, which means components should not have global state. Global state in components is typically just cause for problems. For example it may create ordering issues or unintended side effects and so on.
```

* `reserved` - parameter reserved for future extensibility, currently unused and must be `0`.
* `delegate` - out parameter which receives the native function pointer to the requested managed method.

> It is allowed to call the returned runtime helper many times for different assemblies or just different methods from the same assembly. It is not required to get the helper every time. The implementation of the helper will cache loaded assemblies, so asking to load the same assembly twice will end up loading it only once and reusing it from that point onward. Ideally components should not take a dependency on this behavior, which means components should not have global state. Global state in components is typically just cause for problems. For example it may create ordering issues or unintended side effects and so on.

Reviewer 2: I really wasn't thinking about global state. Of course that plays into it too. I was really thinking to some extent about the shared types.

I think the caching is part of the current implementation contract.

One of the reasons for caching is for an assembly in with multiple entry-points.  Say an XML reader and an XML beautifier and an XML writer....
